geomesa-cassandra/geomesa-cassandra-datastore/src/main/scala/org/locationtech/geomesa/cassandra/index/CassandraFeatureIndex.scala

```diff
@@ -0,0 +1,186 @@
+import org.locationtech.geomesa.index.index.ClientSideFiltering.RowAndValue
+import org.locationtech.geomesa.index.index.{ClientSideFiltering, IndexAdapter}
+import org.locationtech.geomesa.curve.TimePeriod
+import org.opengis.feature.simple.SimpleFeatureType
+import org.opengis.feature.simple.SimpleFeature
+import org.opengis.filter.Filter
+
+import scala.collection.JavaConversions._
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+import java.util.Date
+
+object CassandraFeatureIndex extends CassandraIndexManagerType {
+  // note: keep in priority order for running full table scans
+  override val AllIndices: Seq[CassandraFeatureIndex] =
+    Seq(CassandraZ3Index, CassandraXZ3Index, CassandraZ2Index, CassandraXZ2Index, CassandraIdIndex, CassandraAttributeIndex)
+
+  override val CurrentIndices: Seq[CassandraFeatureIndex] = AllIndices
+}
+
+trait CassandraFeatureIndex extends CassandraFeatureIndexType
+    with IndexAdapter[CassandraDataStore, CassandraFeature, Statement, Statement] with ClientSideFiltering[Row] {
+  import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+
+  type RowToValues = (Array[Byte]) => Any
+
+  var getRowValues:RowToValues = {row => (0L, 0L, "")}
+  var tableName = ""
+  var ks = ""
+
+  def getRowValuesFromSFT(sft: SimpleFeatureType): RowToValues = {
+    // table sharing + shard + 2 byte short + 8 byte long + fid
+    val sharing = sft.getTableSharingPrefix
+    (row) => {
+      var curr = if (sharing == "") 0 else 1
+      // use shard and bin for pk
+      val pk:java.lang.Integer = ((row(curr) & 0xFF) << 16) | ((row(curr + 1) & 0xFF) << 8) | (row(curr + 2) & 0xFF);
+      curr += 3
+      val z = ByteBuffer.wrap(row, curr, 8).getLong
+      curr += 8
+      val fid = new String(row, curr, row.length - curr, StandardCharsets.UTF_8)
+      (pk, z, fid, sharing)
+    }
+  }
+
+  override def configure(sft: SimpleFeatureType, ds: CassandraDataStore): Unit = {
+    super.configure(sft, ds)
+    tableName = getTableName(sft.getTypeName, ds)
+    getRowValues = getRowValuesFromSFT(sft)
+
+    val cluster = ds.session.getCluster
+    ks = ds.session.getLoggedKeyspace
+    val table = cluster.getMetadata().getKeyspace(ks).getTable(tableName);
+
+    if (table == null){
+      ds.session.execute(s"CREATE TABLE $tableName (pk int, sharing text, z bigint, fid text, sf blob, PRIMARY KEY (pk, sharing, z, fid))")
+    }
+  }
+
+  override def delete(sft: SimpleFeatureType, ds: CassandraDataStore, shared: Boolean): Unit = {
+    val name = getTableName(sft.getTypeName, ds)
+    ds.session.execute(s"drop table $tableName")
+  }
+
+  override protected def createInsert(row: Array[Byte], cf: CassandraFeature): Statement = {
+    val (pk:java.lang.Integer, z:java.lang.Long, fid:java.lang.String, sharing:java.lang.String) = getRowValues(row)
+    new SimpleStatement(s"INSERT INTO $tableName (pk, z, fid, sharing, sf) values (?, ?, ?, ?, ?)",
```

Comment:
this might need to be changed so return a string for the statement and the binding values, so that the table name can be inserted after it's available in CassandraFeatureWriter